File: samples/go/encode-perf-rig/string-encoder.go
// Copyright 2016 The Noms Authors. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package main

import (
	"io"
	"math/big"
)

type StringEncoderDecoder struct{}

func NewStringEncodedDecoder() StringEncoderDecoder {
	return StringEncoderDecoder{}
}

func (sed StringEncoderDecoder) Encode(w io.Writer, n *big.Float) error {
	// TODO - big.Float.MarshalText?
	// TODO - big.Float.Append
	str := []byte(n.Text('g', -1))
	_, err := w.Write(str)
	return err
}

func (sed StringEncoderDecoder) Decode(r io.Reader, n *big.Float) error {
	n.SetFloat64(0)
	n.SetPrec(ENCODER_DECODER_PREC)

	buf := make([]byte, 256)
	if _, err := r.Read(buf); err != nil {
		return err
	}
	_, _, err := n.Parse(string(buf), 10)
	return err
}

File: js/perf/encode-perf-rig/src/string-encoder.js
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

// @flow

export class StringEncoderDecoder {

  // write n to buf, return number of bytes written
  encode(buf: Buffer, n: number): number {
    if (n < 1e20) {
      // $FlowIssue: Buffer.prototype.write returns a number
      return buf.write(n.toString(10));
    }
    // $FlowIssue: Buffer.prototype.write returns a number
    return buf.write(n.toExponential());
  }

  // read from buf to return number
  decode(buf: Buffer): number {
    const s = buf.toString();
    return parseFloat(s);
  }
}
